### PR TITLE
Implement version-and-bookmark

### PR DESCRIPTION
**Pull request by Maxwell-Yang-PDFTron**

### CODE REVIEW

File: android/src/main/java/com/pdftron/reactnative/modules/DocumentViewModule.java [hunk 1]
```diff
@@ -31,6 +32,50 @@ public String getName() {
         return REACT_CLASS;
     }
 
+    @ReactMethod
+    public void getVersion(final int tag, final Promise promise) {
```
Comment (sgong-pdftron) [inline android/src/main/java/com/pdftron/reactnative/modules/DocumentViewModule.java:36]: getVersion should be on RNPdfnet

**Review by sgong-pdftron — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/pdftron/reactnative/modules/RNPdftronModule.java [hunk 2]
```diff
@@ -66,4 +66,33 @@ public void encryptDocument(final String filePath, final String password, final
...
         }
     }
+
+    @ReactMethod
+    public void getVersion(final Promise promise) {
+        getReactApplicationContext().runOnUiQueueThread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    double version = PDFNet.getVersion();
```
Comment (sgong-pdftron) [inline android/src/main/java/com/pdftron/reactnative/modules/RNPdftronModule.java:76]: ? repeated on the next line

File: src/DocumentView/DocumentView.js [hunk 3]
```diff
@@ -89,7 +90,9 @@ export default class DocumentView extends PureComponent {
         });
       }
     } else if (event.nativeEvent.onAnnotationChanged) {
+      console.log("1");
```
Comment (sgong-pdftron) [inline src/DocumentView/DocumentView.js:93]: logs

**Review by sgong-pdftron — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/pdftron/reactnative/views/DocumentView.java [hunk 4]
```diff
@@ -1209,6 +1214,64 @@ public void annotationsCouldNotBeAdded(String s) {
         }
     };
 
+    private ToolManager.PdfDocModificationListener mPdfDocModificationListener = new ToolManager.PdfDocModificationListener() {
+        @Override
+        public void onBookmarkModified() {
+            WritableMap params = Arguments.createMap();
+            params.putString(ON_BOOKMARK_CHANGED, ON_BOOKMARK_CHANGED);
+            String bookmarkJson = null;
+            if (getPdfDoc() != null) {
```
Comment (sgong-pdftron) [inline android/src/main/java/com/pdftron/reactnative/views/DocumentView.java:1223]: null check should include the whole block, if no doc, the event should not be triggered

**Review by sgong-pdftron — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/pdftron/reactnative/modules/DocumentViewModule.java [hunk 5]
```diff
@@ -11,6 +11,7 @@
 import com.facebook.react.bridge.ReadableArray;
 import com.facebook.react.bridge.ReadableMap;
 import com.facebook.react.bridge.WritableMap;
+import com.pdftron.pdf.PDFNet;
```
Comment (sgong-pdftron) [inline android/src/main/java/com/pdftron/reactnative/modules/DocumentViewModule.java:14]: cleanup

**Review by sgong-pdftron — CHANGES_REQUESTED** [review]

**Comment by sgong-pdftron** [comment]
resolve conflict

File: src/DocumentView/DocumentView.js [hunk 6]
```diff
@@ -175,6 +182,14 @@ export default class DocumentView extends PureComponent {
     return Promise.resolve();
   }
 
+  importBookmarkJson = (bookmarkJson) => {
+    const tag = findNodeHandle(this._viewerRef);
+    if (tag != null) {
+      DocumentViewManager.importBookmarkJson(tag, bookmarkJson);
+    }
+    return Promise.resolve();
```
Comment (sgong-pdftron) [inline src/DocumentView/DocumentView.js:190]: no return value

**Review by sgong-pdftron — CHANGES_REQUESTED** [review]

**Review by sgong-pdftron — APPROVED** [review]

File: ios/RNPdftron.m [hunk 7]
```diff
@@ -52,6 +52,28 @@ - (dispatch_queue_t)methodQueue
     }
 }
 
+RCT_EXPORT_METHOD(getVersion:(RCTPromiseResolveBlock)resolve
+                 rejecter:(RCTPromiseRejectBlock)reject)
+{
+    @try {
+        resolve([@"PDFNet " stringByAppendingFormat:@"%f", [PTPDFNet GetVersion]]);
```
Comment (JamesTron) [inline ios/RNPdftron.m:59]: Seems to me this should simply return a double. Why prefix PDFNet?

File: ios/RNPdftron.m [hunk 8]
```diff
@@ -52,6 +52,28 @@ - (dispatch_queue_t)methodQueue
...
+    }
+    @catch (NSException *exception) {
+        reject(@"get_failed", @"Failed to get PDFNet version", [self errorFromException:exception]);
+    }
+}
+
+RCT_EXPORT_METHOD(getPlatformVersion:(RCTPromiseResolveBlock)resolve
```
Comment (JamesTron) [inline ios/RNPdftron.m:66]: Is this something PDFTron would be expected to have? Doesn't seem to be within the scope of our library. Why are we adding this?

File: ios/RNTPTDocumentView.m [hunk 9]
```diff
@@ -2052,6 +2063,36 @@ -(NSString*)generateXfdfCommand:(PTVectorAnnot*)added modified:(PTVectorAnnot*)m
...
+    [self.pdfViewCtrl DocLockReadWithBlock:^(PTPDFDoc * _Nullable doc) {
+        json = [PTBookmarkManager.defaultManager exportBookmarksFromDoc:doc];
+    } error:&error];
+    
+    if(error)
+    {
+        NSLog(@"Error: There was an error while trying to export the bookmark json on events triggered. %@", error.localizedDescription);
+    }
+    
+    if ([self.delegate respondsToSelector:@selector(bookmarkChanged:bookmarkJson:)]) {
```
Comment (JamesTron) [inline ios/RNTPTDocumentView.m:2092]: You could do this check around the whole function. If the person is not listening for changes, we can skip the work of acquiring a lock and extracting the string.

**Review by JamesTron — CHANGES_REQUESTED** [review]

File: ios/RNPdftron.m (hunk 8 above)
Comment (Maxwell-Yang-PDFTron) [reply ios/RNPdftron.m:66]: FL has it (though not in README), so I thought it would be nice to have it in RN as well.

File: ios/RNPdftron.m (hunk 7 above)
Comment (Maxwell-Yang-PDFTron) [reply ios/RNPdftron.m:59]: Copied from FL.

https://github.com/PDFTron/pdftron-flutter/blob/master/ios/Classes/PdftronFlutterPlugin.m#L544

Comment (JamesTron) [reply ios/RNPdftron.m:59]: I don't think either should return a string. But it's not worth changing now.

File: ios/RNPdftron.m (hunk 8 above)
Comment (JamesTron) [reply ios/RNPdftron.m:66]: OK

**Review by JamesTron — APPROVED** [review]

**Pull request merged**